<a href="https://colab.research.google.com/github/sergiocostaifes/PPCOMP_DM/blob/main/notebooks/05_feature_engineering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# NB05 — Engenharia de Atributos

## 1. Contexto

Após a construção da série temporal agregada no NB03 e da identificação oficial de
episódios críticos no NB04, este notebook realiza a etapa de engenharia de atributos a
partir da série temporal contínua.

Enquanto o NB04 segmenta a série em janelas críticas e episódios, o papel do NB05 é
transformar essa série em uma base analítica enriquecida com variáveis capazes de
representar dinâmica temporal, intensidade relativa de falhas e padrões recentes de
comportamento do sistema.

## 2. Objetivo

O objetivo deste notebook é:
- construir atributos derivados a partir da série temporal contínua;
- preservar a coerência com a métrica oficial de criticidade baseada em `fail_rate`;
- representar dependência temporal, tendência, volatilidade e variação recente;
- tratar adequadamente janelas vazias sem destruir a continuidade da série;
- preparar a base para as etapas posteriores de rotulagem e modelagem supervisionada.

## 3. Papel no Pipeline Atualizado

Este notebook conecta a segmentação crítica do NB04 com a formulação supervisionada das
etapas seguintes.

Seu papel é produzir uma base temporal enriquecida que:
- preserve a regularidade da série;
- represente o histórico recente de cada janela;
- capture sinais potencialmente úteis para antecipação de comportamentos críticos;
- mantenha rastreabilidade metodológica em relação à definição oficial de episódios.

Assim, o NB05 fornece a base explicativa utilizada na rotulagem temporal e na modelagem
supervisionada posterior.

## 4. Métrica Base

A engenharia de atributos é centrada na métrica oficial do pipeline:

`fail_rate = n_failed / n_events`

Essa escolha mantém coerência conceitual com o NB04, no qual a detecção de janelas
críticas e episódios foi definida a partir dessa mesma taxa.

Nesta etapa, a série enriquecida com `fail_rate` e `is_critical`, produzida no NB04,
passa a ser utilizada como referência principal.

## 5. Tipos de Atributos Construídos

Os atributos construídos incluem:

### 5.1 Atributos estruturais da série
- variáveis já presentes na série agregada, como volume de eventos, falhas, máquinas,
  coleções e contagens por tipo de evento;
- flag de janela vazia (`is_empty_window`) para representar explicitamente buckets sem
  eventos.

### 5.2 Dependência temporal
- lags de curto prazo da taxa de falha:
  - `lag_1`
  - `lag_2`
  - `lag_3`

### 5.3 Tendência local
- média móvel de 1 hora:
  - `rolling_mean_1h`

### 5.4 Volatilidade local
- desvio-padrão móvel de 1 hora:
  - `rolling_std_1h`

### 5.5 Dinâmica recente
- variação percentual entre janelas consecutivas:
  - `pct_change`

### 5.6 Normalização adaptativa e causal
- z-score expansivo baseado apenas no histórico disponível até `t-1`:
  - `zscore_expanding`

Essa escolha substitui uma normalização global estática, evitando uso de estatísticas
calculadas sobre toda a série, inclusive sobre o futuro.

## 6. Tratamento de Janelas Vazias

A série contínua construída no NB03 contém janelas sem eventos, preservadas para manter a
regularidade temporal.

Nesta etapa:
- essas janelas são mantidas;
- recebem `is_empty_window = 1`;
- os campos médios `mean_priority`, `mean_req_cpus` e `mean_req_mem`, quando ausentes,
  são imputados com 0.0 exclusivamente para viabilizar uso analítico consistente.

Essa decisão não altera a contagem de eventos ou falhas e apenas torna a representação
tabular mais estável para modelagem posterior.

## 7. Saídas Esperadas

Ao final deste notebook, espera-se obter:
- dataset enriquecido com atributos derivados;
- base temporal contínua e coerente com a segmentação oficial do NB04;
- preservação dos rótulos estruturais úteis para diagnóstico;
- artefatos persistidos para uso nas próximas etapas;
- resumo quantitativo da engenharia de atributos realizada.

## 8. Relação com as Próximas Etapas

Este notebook alimenta diretamente:
- NB06 — Derivação de estados e formulação supervisionada;
- etapas posteriores de modelagem e avaliação.

A qualidade das features construídas impacta diretamente a capacidade dos modelos de
representar padrões temporais associados à ocorrência de estados críticos futuros.

## 9. Observações Metodológicas

Todas as features temporais construídas nesta etapa respeitam causalidade:
- não utilizam informação de janelas futuras;
- são derivadas exclusivamente da janela atual e do histórico disponível.

Além disso:
- a definição oficial de criticidade herdada do NB04 é preservada;
- janelas vazias permanecem explicitamente representadas na série;
- a normalização estatística deixa de ser global e passa a ser histórica e causal.

Este notebook estabelece a base explicativa que será utilizada nas etapas supervisionadas
do pipeline.

In [1]:
# ============================================================
# 05_feature_engineering.ipynb
# Engenharia de atributos (5-min)
# Métrica base oficial: fail_rate = n_failed / n_events
# Pipeline PPCOMP_DM (Google Cluster Trace)
# ============================================================

# -----------------------------
# 0) BOOTSTRAP (Colab + Repo)
# -----------------------------
from pathlib import Path
import os
import sys
import subprocess
import importlib
import random
import json

import numpy as np
import pandas as pd
from IPython.display import display

SEED = 42
random.seed(SEED)
np.random.seed(SEED)

# Mount Google Drive
if not Path("/content/drive/MyDrive").exists():
    from google.colab import drive
    drive.mount("/content/drive")
else:
    print("[Bootstrap] Google Drive já montado.")

REPO_DIR = Path("/content/drive/MyDrive/Mestrado/PPCOMP_DM")
GITHUB_REPO = "https://github.com/sergiocostaifes/PPCOMP_DM.git"

if not REPO_DIR.exists():
    REPO_DIR.parent.mkdir(parents=True, exist_ok=True)
    print(f"[Bootstrap] Clonando repositório em: {REPO_DIR}")
    subprocess.run(["git", "clone", GITHUB_REPO, str(REPO_DIR)], check=True)
else:
    try:
        print("[Bootstrap] Atualizando repositório (git pull)...")
        subprocess.run(["git", "-C", str(REPO_DIR), "pull"], check=True)
    except Exception as e:
        print("[Bootstrap] Aviso: não foi possível atualizar:", e)

os.chdir(str(REPO_DIR))
print("[Bootstrap] CWD =", os.getcwd())

repo_str = str(REPO_DIR)
if repo_str not in sys.path:
    sys.path.insert(0, repo_str)

importlib.invalidate_caches()

from src.paths import FEATURES_PATH, REPORTS_PATH, ensure_dirs
ensure_dirs()

print("FEATURES_PATH =", FEATURES_PATH)
print("REPORTS_PATH =", REPORTS_PATH)

def log(msg: str) -> None:
    print(f"[05_feature_engineering] {msg}")

# -----------------------------
# 1) Metadados do NB04
# -----------------------------
NB04_SUMMARY_FILE = REPORTS_PATH / "04_detect_episodes_summary.json"

nb04_summary = {}
if NB04_SUMMARY_FILE.exists():
    nb04_summary = json.loads(NB04_SUMMARY_FILE.read_text(encoding="utf-8"))
    log(f"Resumo do NB04 carregado: {NB04_SUMMARY_FILE}")
else:
    log("Resumo do NB04 não encontrado; seguindo sem metadados adicionais.")

SCENARIO_LABEL = nb04_summary.get("scenario_label", "unknown")
REMOVE_HOUR_ZERO = nb04_summary.get("remove_hour_zero", None)
EXPECTED_CRITICAL_WINDOWS = nb04_summary.get("critical_windows", None)
EXPECTED_EPISODES = nb04_summary.get("episodes_detected", None)

# -----------------------------
# 2) Leitura da série scored + episódios
# -----------------------------
SERIES_FILE = FEATURES_PATH / "window_5min_series_scored.parquet"
EPISODES_FILE = FEATURES_PATH / "episodes_detected.parquet"

assert SERIES_FILE.exists(), f"Arquivo não encontrado: {SERIES_FILE}"
assert EPISODES_FILE.exists(), f"Arquivo não encontrado: {EPISODES_FILE}"

df = pd.read_parquet(SERIES_FILE)
episodes = pd.read_parquet(EPISODES_FILE)

df = df.sort_values("bucket_id").reset_index(drop=True)
episodes = episodes.sort_values(["start_bucket", "end_bucket"]).reset_index(drop=True)

log(f"Shape série scored: {df.shape}")
log(f"Shape episodes: {episodes.shape}")

required_series_cols = {
    "bucket_id",
    "bucket_start_us",
    "n_events",
    "n_failed",
    "fail_rate",
    "is_critical",
}
missing_series = required_series_cols - set(df.columns)
assert not missing_series, f"Colunas ausentes em series_scored: {missing_series}"

required_episode_cols = {"start_bucket", "end_bucket", "duration_windows"}
missing_ep = required_episode_cols - set(episodes.columns)
assert not missing_ep, f"Colunas ausentes em episodes_detected: {missing_ep}"

assert df["bucket_id"].is_monotonic_increasing, "bucket_id não está ordenado."
assert df["bucket_id"].is_unique, "Foram encontrados bucket_id duplicados."

bucket_diff = df["bucket_id"].diff().dropna()
n_internal_gaps_gt1 = int((bucket_diff > 1).sum())
assert n_internal_gaps_gt1 == 0, "A série scored deveria estar contínua após o NB04."

# -----------------------------
# 3) Validações de consistência com o NB04
# -----------------------------
critical_windows_input = int(df["is_critical"].sum())
episodes_detected_input = int(len(episodes))
critical_windows_from_episodes = int(episodes["duration_windows"].sum())

log(f"Critical windows (input scored): {critical_windows_input}")
log(f"Episodes detected (input): {episodes_detected_input}")
log(f"Critical windows (sum duration_windows): {critical_windows_from_episodes}")

assert critical_windows_input == critical_windows_from_episodes, (
    "Inconsistência entre is_critical da série e soma de duration_windows dos episódios."
)

if EXPECTED_CRITICAL_WINDOWS is not None:
    assert critical_windows_input == int(EXPECTED_CRITICAL_WINDOWS), (
        f"Total de janelas críticas difere do NB04: "
        f"esperado={EXPECTED_CRITICAL_WINDOWS}, observado={critical_windows_input}"
    )

if EXPECTED_EPISODES is not None:
    assert episodes_detected_input == int(EXPECTED_EPISODES), (
        f"Total de episódios difere do NB04: "
        f"esperado={EXPECTED_EPISODES}, observado={episodes_detected_input}"
    )

# -----------------------------
# 4) Tratamento de janelas vazias e imputações locais
# -----------------------------
df["is_empty_window"] = (df["n_events"] == 0).astype("int8")
empty_windows = int(df["is_empty_window"].sum())

mean_cols_to_impute = [
    c for c in ["mean_priority", "mean_req_cpus", "mean_req_mem"]
    if c in df.columns
]

na_before_impute = {c: int(df[c].isna().sum()) for c in mean_cols_to_impute}

for c in mean_cols_to_impute:
    df[c] = df[c].fillna(0.0).astype("float32")

na_after_impute = {c: int(df[c].isna().sum()) for c in mean_cols_to_impute}

log(f"Janelas vazias: {empty_windows}")
log(f"NaNs antes da imputação: {na_before_impute}")
log(f"NaNs após a imputação : {na_after_impute}")

# -----------------------------
# 5) Construção de features temporais causais
# -----------------------------
BASE = "fail_rate"

# Lags curtos
df["lag_1"] = df[BASE].shift(1)
df["lag_2"] = df[BASE].shift(2)
df["lag_3"] = df[BASE].shift(3)

# Rolling 1 hora (12 janelas de 5 min), incluindo a janela atual
W = 12
df["rolling_mean_1h"] = df[BASE].rolling(W, min_periods=1).mean().astype("float32")
df["rolling_std_1h"] = (
    df[BASE].rolling(W, min_periods=1).std(ddof=0).fillna(0.0).astype("float32")
)

# Variação percentual entre janelas consecutivas
df["pct_change"] = df[BASE].pct_change()
df["pct_change"] = (
    df["pct_change"]
    .replace([np.inf, -np.inf], np.nan)
    .fillna(0.0)
    .astype("float32")
)

# Normalização adaptativa e causal:
# z-score baseado apenas no histórico disponível até t-1
hist_mean = df[BASE].shift(1).expanding(min_periods=1).mean()
hist_std = df[BASE].shift(1).expanding(min_periods=2).std(ddof=0)

df["zscore_expanding"] = (
    ((df[BASE] - hist_mean) / hist_std.replace(0, np.nan))
    .replace([np.inf, -np.inf], np.nan)
    .fillna(0.0)
    .astype("float32")
)

# -----------------------------
# 6) Limpeza final: remover apenas linhas sem lags suficientes
# -----------------------------
rows_before_trim = int(len(df))

lag_cols = ["lag_1", "lag_2", "lag_3"]
rows_missing_lag = int(df[lag_cols].isna().any(axis=1).sum())

df = df.dropna(subset=lag_cols).reset_index(drop=True)

rows_after_trim = int(len(df))
rows_removed_trim = rows_before_trim - rows_after_trim

log(f"Rows antes do trim por lags: {rows_before_trim}")
log(f"Rows removidas por lags    : {rows_removed_trim}")
log(f"Rows finais                : {rows_after_trim}")

# Após o trim, converter lags para float32
for c in lag_cols:
    df[c] = df[c].astype("float32")

# -----------------------------
# 7) Diagnósticos finais
# -----------------------------
critical_windows_output = int(df["is_critical"].sum())
critical_ratio_output = float(df["is_critical"].mean())

feature_cols_created = [
    "is_empty_window",
    "lag_1",
    "lag_2",
    "lag_3",
    "rolling_mean_1h",
    "rolling_std_1h",
    "pct_change",
    "zscore_expanding",
]

numeric_cols_final = df.select_dtypes(include=[np.number]).columns.tolist()

log(f"Critical windows após trim: {critical_windows_output}")
log(f"Critical ratio após trim   : {critical_ratio_output:.6f}")
log(f"Shape final               : {df.shape}")

# -----------------------------
# 8) Persistência
# -----------------------------
OUT_FILE = FEATURES_PATH / "window_5min_features.parquet"
OUT_SCENARIO_FILE = FEATURES_PATH / f"window_5min_features_{SCENARIO_LABEL}.parquet"

df.to_parquet(OUT_FILE, compression="snappy", index=False)
df.to_parquet(OUT_SCENARIO_FILE, compression="snappy", index=False)

log(f"Features salvas (canônico): {OUT_FILE}")
log(f"Features salvas (cenário): {OUT_SCENARIO_FILE}")

# -----------------------------
# 9) Summary JSON
# -----------------------------
summary = {
    "scenario_label": SCENARIO_LABEL,
    "remove_hour_zero": REMOVE_HOUR_ZERO,
    "input_series_file": str(SERIES_FILE),
    "input_episodes_file": str(EPISODES_FILE),

    "rows_input": int(rows_before_trim),
    "rows_output": int(rows_after_trim),
    "rows_removed_trim": int(rows_removed_trim),

    "critical_windows_input": int(critical_windows_input),
    "critical_windows_output": int(critical_windows_output),
    "episodes_detected_input": int(episodes_detected_input),

    "empty_windows": int(empty_windows),
    "na_before_impute": na_before_impute,
    "na_after_impute": na_after_impute,

    "base_metric": BASE,
    "feature_cols_created": feature_cols_created,
    "numeric_cols_final": numeric_cols_final,
    "n_numeric_cols_final": int(len(numeric_cols_final)),

    "critical_ratio_output": float(critical_ratio_output),

    "output_file_canonical": str(OUT_FILE),
    "output_file_scenario": str(OUT_SCENARIO_FILE),
}

SUMMARY_FILE = REPORTS_PATH / "05_feature_engineering_summary.json"
SUMMARY_SCENARIO_FILE = REPORTS_PATH / f"05_feature_engineering_summary_{SCENARIO_LABEL}.json"

SUMMARY_FILE.write_text(json.dumps(summary, indent=2, ensure_ascii=False), encoding="utf-8")
SUMMARY_SCENARIO_FILE.write_text(json.dumps(summary, indent=2, ensure_ascii=False), encoding="utf-8")

log(f"Resumo salvo (canônico): {SUMMARY_FILE}")
log(f"Resumo salvo (cenário): {SUMMARY_SCENARIO_FILE}")

# -----------------------------
# 10) Visualização rápida
# -----------------------------
print("\n=== RESUMO RÁPIDO ===")
print(f"Scenario label            : {SCENARIO_LABEL}")
print(f"Remove hour 0             : {REMOVE_HOUR_ZERO}")
print(f"Rows input                : {rows_before_trim}")
print(f"Rows output               : {rows_after_trim}")
print(f"Rows removed by trim      : {rows_removed_trim}")
print(f"Critical windows input    : {critical_windows_input}")
print(f"Critical windows output   : {critical_windows_output}")
print(f"Episodes detected input   : {episodes_detected_input}")
print(f"Empty windows             : {empty_windows}")
print(f"Critical ratio output     : {critical_ratio_output:.6f}")
print(f"Feature cols created      : {feature_cols_created}")
print(f"n_numeric_cols_final      : {len(numeric_cols_final)}")

print("\n=== HEAD FEATURES ===")
display(df.head())

Mounted at /content/drive
[Bootstrap] Atualizando repositório (git pull)...
[Bootstrap] CWD = /content/drive/MyDrive/Mestrado/PPCOMP_DM
FEATURES_PATH = /content/drive/MyDrive/Mestrado/02-datasets/03-features
REPORTS_PATH = /content/drive/MyDrive/Mestrado/04-reports
[05_feature_engineering] Resumo do NB04 carregado: /content/drive/MyDrive/Mestrado/04-reports/04_detect_episodes_summary.json
[05_feature_engineering] Shape série scored: (8930, 20)
[05_feature_engineering] Shape episodes: (285, 17)
[05_feature_engineering] Critical windows (input scored): 390
[05_feature_engineering] Episodes detected (input): 285
[05_feature_engineering] Critical windows (sum duration_windows): 390
[05_feature_engineering] Janelas vazias: 2
[05_feature_engineering] NaNs antes da imputação: {'mean_priority': 2, 'mean_req_cpus': 2, 'mean_req_mem': 2}
[05_feature_engineering] NaNs após a imputação : {'mean_priority': 0, 'mean_req_cpus': 0, 'mean_req_mem': 0}
[05_feature_engineering] Rows antes do trim por lag

,bucket_id,bucket_start_us,n_events,n_failed,n_machines,n_collections,mean_priority,mean_req_cpus,mean_req_mem,req_cpus_presence_rate,...,fail_rate,is_critical,is_empty_window,lag_1,lag_2,lag_3,rolling_mean_1h,rolling_std_1h,pct_change,zscore_expanding
0,3,900000000,32,2,32,18,135.656250,0.009495,0.003928,1.0,...,0.062500,0,0,0.321429,0.000000,0.656877,0.260201,0.258791,-0.805556,-0.982895
1,4,1200000000,29,6,28,20,233.068970,0.012932,0.003717,1.0,...,0.206897,0,0,0.062500,0.321429,0.000000,0.249540,0.232449,2.310345,-0.205976
2,5,1500000000,25,5,25,19,253.360001,0.011083,0.012531,1.0,...,0.200000,0,0,0.206897,0.062500,0.321429,0.241284,0.212998,-0.033333,-0.213123
3,6,1800000000,25,4,25,21,241.559998,0.011335,0.007542,1.0,...,0.160000,0,0,0.200000,0.206897,0.062500,0.229672,0.199239,-0.200000,-0.381617
4,7,2100000000,40,12,37,25,249.399994,0.016829,0.006852,1.0,...,0.300000,0,0,0.160000,0.200000,0.206897,0.238463,0.187816,0.875000,0.352986


## 10. Conclusão da Etapa

A etapa de engenharia de atributos foi concluída com sucesso, resultando em uma base
temporal enriquecida e coerente com a segmentação crítica oficial definida no NB04.

Nesta execução, o notebook utilizou como entradas:
- a série contínua enriquecida com `fail_rate` e `is_critical`;
- o conjunto oficial de episódios detectados.

A base resultante:
- preserva a continuidade temporal da série;
- mantém coerência com a métrica oficial de criticidade baseada em `fail_rate`;
- incorpora atributos derivados capazes de representar histórico recente, tendência,
  volatilidade, variação e normalização adaptativa;
- trata explicitamente as janelas vazias da série contínua.

O dataset final foi salvo em:
- `window_5min_features.parquet`
- `window_5min_features_keep_hour0.parquet`

## 11. Achados Experimentais da Execução Atual

A execução atual consolidou a nova formulação da etapa de atributos, tornando-a mais
causal, mais estável e mais alinhada ao pipeline revisado.

### 11.1 Consistência com o NB04

A leitura dos artefatos do NB04 confirmou:
- `critical_windows_input = 390`
- `episodes_detected_input = 285`
- soma das `duration_windows` dos episódios = 390

Esses valores mostram que a série scored e os episódios detectados estão consistentes entre
si e com a saída oficial da etapa anterior.

### 11.2 Tratamento das janelas vazias

A série de entrada continha:
- `empty_windows = 2`

Nessas janelas, os campos médios estruturais apresentavam valores ausentes:
- `mean_priority`: 2 valores ausentes
- `mean_req_cpus`: 2 valores ausentes
- `mean_req_mem`: 2 valores ausentes

Após a imputação local:
- todos esses campos passaram a ter 0 valores ausentes

Essa decisão foi adotada apenas para estabilizar a representação tabular da série contínua,
sem alterar as contagens de eventos ou falhas e sem romper a regularidade temporal.

### 11.3 Features construídas

Nesta execução, foram criadas explicitamente as seguintes features derivadas:
- `is_empty_window`
- `lag_1`
- `lag_2`
- `lag_3`
- `rolling_mean_1h`
- `rolling_std_1h`
- `pct_change`
- `zscore_expanding`

Essas variáveis representam:
- presença de janela vazia;
- memória de curto prazo;
- tendência local em 1 hora;
- volatilidade local em 1 hora;
- variação recente entre janelas;
- normalização adaptativa baseada apenas no histórico disponível.

### 11.4 Substituição da normalização global

Um dos avanços metodológicos desta versão foi a substituição do antigo `zscore_global`
por `zscore_expanding`.

Isso significa que a normalização:
- deixa de usar média e desvio calculados sobre toda a série;
- passa a usar apenas o histórico disponível até `t-1`;
- evita incorporar estatísticas do futuro na própria representação da janela atual.

Esse ajuste fortalece a alegação de causalidade temporal da engenharia de atributos.

### 11.5 Trim por lags

Antes da remoção das linhas sem histórico suficiente, a base possuía:
- `rows_input = 8930`

Foram removidas:
- `rows_removed_trim = 3`

O dataset final ficou com:
- `rows_output = 8927`

Essa remoção corresponde exclusivamente à indisponibilidade dos três lags de curto prazo
nas primeiras janelas da série, e não a perdas decorrentes de `NaN` estruturais em outras
variáveis.

### 11.6 Janelas críticas após o trim

A base de entrada continha:
- `critical_windows_input = 390`

Após o trim dos lags:
- `critical_windows_output = 389`

Essa diferença de uma janela crítica decorre da remoção das primeiras linhas da série
por insuficiência de histórico para os lags, e não de qualquer mudança na regra de
criticidade herdada do NB04.

A razão final de criticidade na base produzida foi:
- `critical_ratio_output = 0.043576`

### 11.7 Estrutura final da base

A base final produzida nesta etapa apresentou:
- 8.927 linhas
- 28 colunas numéricas

As colunas numéricas finais incluem:
- variáveis estruturais da série agregada;
- contagens por tipo de evento;
- `fail_rate` e `is_critical`;
- `is_empty_window`;
- lags;
- estatísticas móveis;
- variação percentual;
- normalização expansiva causal.

Essa estrutura fornece uma base rica e consistente para a etapa seguinte de rotulagem e
formulação supervisionada.

## 12. Importância para o Pipeline Analítico

O NB05 estabelece a base explicativa do pipeline.

Seu papel não é definir criticidade — isso já foi feito no NB04 —, mas transformar a
série contínua e segmentada em um conjunto de variáveis capazes de representar padrões
temporais úteis para as etapas supervisionadas.

Nesta execução, o notebook fortalece o pipeline em quatro pontos principais:
- preserva coerência com a definição oficial de episódios;
- mantém as janelas vazias explicitamente representadas;
- elimina dependência de normalização global estática;
- restringe o descarte de linhas apenas ao mínimo necessário por causa dos lags.

Esses aspectos tornam a base mais defensável metodologicamente e mais adequada para
modelagem causal.

## 13. Artefatos Gerados

Nesta execução, foram gerados os seguintes artefatos:
- `window_5min_features.parquet`
- `window_5min_features_keep_hour0.parquet`
- `05_feature_engineering_summary.json`
- `05_feature_engineering_summary_keep_hour0.json`

Esses arquivos passam a constituir a referência oficial da etapa de engenharia de atributos
para o cenário `keep_hour0`.

## 14. Encaminhamento do Pipeline

Com a base de atributos devidamente construída e persistida, o pipeline segue para:
- NB06 — Derivação de estados e formulação supervisionada;
- etapas posteriores de modelagem e avaliação.

A partir deste ponto, a formulação supervisionada passa a operar sobre uma base temporal
enriquecida, contínua e alinhada à segmentação crítica oficial do pipeline.